tests/unit/h/views/oidc_test.py

```diff
@@ -74,158 +86,256 @@ def routes(self, pyramid_config):
+        self, user_service, views
     ):
-        OAuth2RedirectSchema.validate.side_effect = ValidationError()
+        user_service.fetch_by_identity.return_value = sentinel.other_user
 
-        with pytest.raises(views.ValidationError) as exc_info:
-            views.ORCIDRedirectViews(pyramid_request).redirect()
+        with pytest.raises(UserConflictError):
+            views.redirect()
+
+    @pytest.mark.usefixtures("with_connect_action")
+    @pytest.mark.usefixtures(
+        "assert_state_removed_from_session",
+        "assert_success_message_was_flashed",
+        "assert_user_was_not_logged_in",
+    )
+    def test_redirect_when_action_connect_and_account_not_yet_connected(
+        self, pyramid_request, orcid_client_service, user_service, user, views, matchers
+    ):
+        user_service.fetch_by_identity.return_value = None
 
-        orcid_client_service.add_identity.assert_not_called()
-        assert exc_info.value == OAuth2RedirectSchema.validate.side_effect
+        result = views.redirect()
 
-    def test_when_token_validation_error(self, pyramid_request, orcid_client_service):
-        orcid_client_service.get_orcid.side_effect = TokenValidationError
+        orcid_client_service.add_identity.assert_called_once_with(
+            user, sentinel.orcid_id
+        )
+        assert result == matchers.Redirect302To(pyramid_request.route_url("account"))
+
+    @pytest.mark.usefixtures(
+        "with_connect_action",
+        "assert_state_removed_from_session",
+        "assert_success_message_was_flashed",
+        "assert_no_account_connection_was_added",
+        "assert_user_was_not_logged_in",
+    )
+    def test_redirect_when_action_connect_and_account_already_connected(
+        self, pyramid_request, user_service, views, matchers
+    ):
+        user_service.fetch_by_identity.return_value = pyramid_request.user
 
-        with pytest.raises(TokenValidationError):
-            views.ORCIDRedirectViews(pyramid_request).redirect()
+        result = views.redirect()
 
-        orcid_client_service.add_identity.assert_not_called()
+        assert result == matchers.Redirect302To(pyramid_request.route_url("account"))
 
-    def test_when_external_request_error(self, pyramid_request, orcid_client_service):
-        orcid_client_service.get_orcid.side_effect = ExternalRequestError
+    @pytest.mark.usefixtures(
+        "with_login_action",
+        "assert_state_removed_from_session",
+        "assert_no_account_connection_was_added",
+        "assert_user_was_not_logged_in",
+        "assert_no_success_message_was_flashed",
+    )
+    def test_redirect_when_action_login_and_no_connected_user_exists(
+        self, views, user_service
+    ):
+        user_service.fetch_by_identity.return_value = None
 
-        with pytest.raises(ExternalRequestError):
-            views.ORCIDRedirectViews(pyramid_request).redirect()
+        with pytest.raises(RuntimeError):
```

Comment:
Maybe add the message here to make it clear this is a temporary thing.

Reply:
:+1: